mpas_analysis/shared/generalized_reader/generalized_reader.py

```diff
@@ -0,0 +1,342 @@
+from ..timekeeping.utility import stringToDatetime, clampToNumpyDatetime64
+
+
+def open_multifile_dataset(file_names, calendar, time_variable_name='Time',
+                           variable_list=None, sel_values=None,
+                           isel_values=None, variable_map=None,
+                           start_date=None, end_date=None,
+                           year_offset=0):  # {{{
+    """
+    Opens and returns an xarray data set given file name(s) and the MPAS
+    calendar name.
+
+    Parameters
+    ----------
+    file_names : list of strings
+        A lsit of file paths to read
+
+    calendar : {'gregorian', 'gregorian_noleap'}, optional
+        The name of one of the calendars supported by MPAS cores
+
+    time_variable_name : string, optional
+        The name of the time variable (typically 'Time' if using a variable_map
+        or 'xtime' if not using a variable_map)
+
+    variable_list : list of strings, optional
+        If present, a list of variables to be included in the data set
+
+    sel_values : dict, optional
+        A dictionary of coordinate names (keys) and values or arrays of
+        values used to slice the variales in the data set.  See
+        xarray.DataSet.sel() for details on how this dictonary is used.
+        An example:
+            selectCorrdValues = {'cellLon': 180.0}
+
+    isel_values : dict, optional
+        A dictionary of coordinate names (keys) and indices, slices or
+        arrays of indices used to slice the variales in the data set.  See
+        xarray.DataSet.isel() for details on how this dictonary is used.
+        An example:
+            isel_values = {'nVertLevels': slice(0, 3),
+                           'nCells': cellIDs}
+
+    variable_map : dict, optional
+        A dictionary with keys that are variable names used by
+        MPAS-Analysis and values that are lists of possible names for the same
+        variable in the MPAS dycore that produced the data set (which may
+        differ between versions).
+
+    start_date, end_date : string or datetime.datetime, optional
+        If present, the first and last dates to be used in the data set.  The
+        time variable is sliced to only include dates within this range.
+
+    year_offset : float, optional
+        An offset used to convert an MPAS date to a date in the range supported
+        by xarray (numpy.datetime64).  Resulting dates must be between 1678
+        and 2622.
+
+    Returns
+    -------
+    ds : An xarray data set.
+
+    Raises
+    ------
+    TypeError
+        If the time variable has an unsupported type (not a date string,
+        a floating-pont number of days since the start of the simulation
+        or a numpy.datatime64 object).
+
+    ValueError
+        If the time variable is not found in the data set.
+
+    Author
+    ------
+    Xylar Asay-Davis
+
+    Last modified
+    -------------
+    02/15/2017
+    """
+
+    def preprocess_fn(ds):
```

Comment:
You can use a partial function here, which would be cleaner because you are essentially doing this by hand.  See https://docs.python.org/2/library/functools.html#functools.partial